ccl/serverccl: TestServerControllerHTTP failed

ccl/serverccl.TestServerControllerHTTP [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8933299?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8933299?buildTab=artifacts#/) on master @ [14b43be03c1c246765be17aa15245493a42c401d](https://github.com/cockroachdb/cockroach/commits/14b43be03c1c246765be17aa15245493a42c401d):


```
=== RUN   TestServerControllerHTTP
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestServerControllerHTTP452427513
    test_log_scope.go:79: use -show-logs to present logs inline
    server_controller_test.go:67: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3439/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/serverccl/serverccl_test_/serverccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/serverccl/server_controller_test.go:67
        	Error:      	Received unexpected error:
        	            	server stop before start
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*serverController).startControlledServer.func3.1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/server_controller_orchestration.go:257
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*serverController).startControlledServer.func3
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/server_controller_orchestration.go:306
        	            	  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) server stop before start
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestServerControllerHTTP
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestServerControllerHTTP452427513
--- FAIL: TestServerControllerHTTP (3.77s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerControllerHTTP.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/serverccl.TestServerControllerHTTP [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9088830?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9088830?buildTab=artifacts#/) on master @ [d68f5cdfb25476aaf874bff0bb6515eba02efcd2](https://github.com/cockroachdb/cockroach/commits/d68f5cdfb25476aaf874bff0bb6515eba02efcd2):


```
=== RUN   TestServerControllerHTTP
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestServerControllerHTTP1953071424
    test_log_scope.go:79: use -show-logs to present logs inline
    server_controller_test.go:67: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3667/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/serverccl/serverccl_test_/serverccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/serverccl/server_controller_test.go:67
        	Error:      	Received unexpected error:
        	            	server stop before start
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*serverController).startControlledServer.func3.1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/server_controller_orchestration.go:257
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*serverController).startControlledServer.func3
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/server_controller_orchestration.go:306
        	            	  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) server stop before start
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestServerControllerHTTP
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestServerControllerHTTP1953071424
--- FAIL: TestServerControllerHTTP (4.32s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerControllerHTTP.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/serverccl/BUILD.bazel
+++ b/pkg/ccl/serverccl/BUILD.bazel
@@ -74,6 +74,7 @@ go_test(
         "//pkg/sql/tests",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
+        "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/testcluster",
         "//pkg/ts/catalog",

--- a/pkg/ccl/serverccl/server_controller_test.go
+++ b/pkg/ccl/serverccl/server_controller_test.go
@@ -25,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/catconstants"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/httputil"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -38,6 +39,8 @@ func TestServerControllerHTTP(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	skip.UnderDeadlock(t, "test triggers many goroutines, which results in conn timeouts and test failures under deadlock")
+
 	ctx := context.Background()
 
 	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{
@@ -45,11 +48,15 @@ func TestServerControllerHTTP(t *testing.T) {
 	})
 	defer s.Stopper().Stop(ctx)
 
+	t.Logf("waking up HTTP server")
+
 	// Retrieve a privileged HTTP client. NB: this also populates
 	// system.web_sessions.
 	aurl := s.AdminURL()
 	client, err := s.GetAdminHTTPClient()
 	require.NoError(t, err)
+
+	t.Logf("retrieving web session details")
 
 	// Now retrieve the entry in the system tenant's web sessions.
 	row := db.QueryRow(`SELECT id,"hashedSecret",username,"createdAt","expiresAt" FROM system.web_sessions`)
@@ -59,12 +66,16 @@ func TestServerControllerHTTP(t *testing.T) {
 	var created, expires time.Time
 	require.NoError(t, row.Scan(&id, &secret, &username, &created, &expires))
 
+	t.Logf("waking up a test tenant")
+
 	// Create our own test tenant with a known name.
 	_, _, err = s.(*server.TestServer).StartSharedProcessTenant(ctx,
 		base.TestSharedProcessTenantArgs{
 			TenantName: "hello",
 		})
 	require.NoError(t, err)
+
+	t.Logf("connecting to the test tenant")
 
 	// Get a SQL connection to the test tenant.
 	sqlAddr := s.ServingSQLAddr()
@@ -75,6 +86,8 @@ func TestServerControllerHTTP(t *testing.T) {
 
 	// This actually uses the connection.
 	require.NoError(t, db2.Ping())
+
+	t.Logf("creating a test user and session")
 
 	// Instantiate the HTTP test username and privileges into the test tenant.
 	_, err = db2.Exec(fmt.Sprintf(`CREATE USER %s`, lexbase.EscapeSQLIdent(username)))
@@ -87,6 +100,8 @@ func TestServerControllerHTTP(t *testing.T) {
 VALUES($1, $2, $3, $4, $5, (SELECT user_id FROM system.users WHERE username = $3))`,
 		id, secret, username, created, expires)
 	require.NoError(t, err)
+
+	t.Logf("configuring the test connections")
 
 	// From this point, we are expecting the ability to access both tenants using
 	// the same cookie jar.
@@ -128,6 +143,8 @@ VALUES($1, $2, $3, $4, $5, (SELECT user_id FROM system.users WHERE username = $3
 		return req
 	}
 
+	t.Logf("retrieving session list from system tenant")
+
 	// Retrieve the session list for the system tenant.
 	req := newreq()
 	req.Header.Set(server.TenantSelectHeader, catconstants.SystemTenantName)
@@ -137,6 +154,8 @@ VALUES($1, $2, $3, $4, $5, (SELECT user_id FROM system.users WHERE username = $3
 	require.Equal(t, len(body.Sessions), 1)
 	require.Equal(t, body.Sessions[0].ApplicationName, "hello system")
 
+	t.Logf("retrieving session list from test tenant")
+
 	// Ditto for the test tenant.
 	req = newreq()
 	req.Header.Set(server.TenantSelectHeader, "hello")
@@ -145,6 +164,8 @@ VALUES($1, $2, $3, $4, $5, (SELECT user_id FROM system.users WHERE username = $3
 	t.Logf("response 2:\n%#v", body)
 	require.Equal(t, len(body.Sessions), 1)
 	require.Equal(t, body.Sessions[0].ApplicationName, "hello hello")
+
+	t.Logf("retrieving session list from system tenant via cookie")
 
 	c := &http.Cookie{
 		Name:     server.TenantSelectCookieName,
@@ -164,6 +185,8 @@ VALUES($1, $2, $3, $4, $5, (SELECT user_id FROM system.users WHERE username = $3
 	require.Equal(t, len(body.Sessions), 1)
 	require.Equal(t, body.Sessions[0].ApplicationName, "hello system")
 
+	t.Logf("retrieving session list from test tenant via cookie")
+
 	c.Value = "hello"
 	client.Jar.SetCookies(purl, []*http.Cookie{c})
 	req = newreq()
@@ -172,6 +195,8 @@ VALUES($1, $2, $3, $4, $5, (SELECT user_id FROM system.users WHERE username = $3
 	t.Logf("response 4:\n%#v", body)
 	require.Equal(t, len(body.Sessions), 1)
 	require.Equal(t, body.Sessions[0].ApplicationName, "hello hello")
+
+	t.Logf("retrieving session list from test tenant via cookie and header")
 
 	// Finally, do it again with both cookie and header. Verify
 	// that the header wins.
@@ -182,6 +207,8 @@ VALUES($1, $2, $3, $4, $5, (SELECT user_id FROM system.users WHERE username = $3
 	t.Logf("response 5:\n%#v", body)
 	require.Equal(t, len(body.Sessions), 1)
 	require.Equal(t, body.Sessions[0].ApplicationName, "hello system")
+
+	t.Logf("end of test")
 }
 
 // TestServerControllerBadHTTPCookies tests the controller's proxy

